sources/packages/backend/apps/queue-consumers/src/services/workflow/workflow-enqueuer.service.ts

```diff
@@ -0,0 +1,119 @@
+ * Manages the operations to search assessments that requires some
+ * workflow operations, for instance, triggering their start or cancellation.
+ */
+@Injectable()
+export class WorkflowEnqueuerService {
+  constructor(
+    private readonly dataSource: DataSource,
+    private readonly applicationService: ApplicationService,
+    @InjectQueue(QueueNames.StartApplicationAssessment)
+    private readonly startAssessmentQueue: Queue<StartAssessmentQueueInDTO>,
+  ) {}
+
+  /**
+   * Search applications with pending assessments to be processed by the assessment workflow.
+   * If no other assessment is being processed for that application the oldest pending
+   * assessment will be queue for start.
+   * @returns process summary log.
+   */
+  async enqueueStartAssessmentWorkflows(): Promise<ProcessSummaryResult> {
+    const result = new ProcessSummaryResult();
+    try {
+      result.summary.push(
+        "Checking database for applications with assessments waiting to be triggered.",
+      );
+      const applications =
+        await this.applicationService.getApplicationsToStartAssessments();
+      result.summary.push(`Found ${applications.length} applications.`);
+      if (!applications.length) {
+        result.summary.push("No applications found");
+        return result;
+      }
+      result.children = await processInParallel(
+        (application: Application) => this.queueNextAssessment(application),
+        applications,
+      );
+    } catch (error: unknown) {
+      result.errors.push(
+        `Error while enqueueing assessment workflows to be processed. ${parseJSONError(
+          error,
+        )}`,
+      );
+    }
+    return result;
+  }
+
+  /**
+   * Queue the next pending assessment for an application.
+   * @param application application with pending assessments.
+   * @returns process summary.
+   */
+  private async queueNextAssessment(
+    application: Application,
+  ): Promise<ProcessSummaryResult> {
+    const result = new ProcessSummaryResult();
+    try {
+      result.summary.push(
+        `Queueing next pending assessment for application id ${application.id}.`,
+      );
+      const [nextAssessment] = application.studentAssessments;
+      result.summary.push(
+        `Found ${application.studentAssessments.length} pending assessment(s). Queueing assessment ${nextAssessment.id}.`,
+      );
+      // Update application and student assessment.
+      await this.dataSource.transaction(async (entityManager) => {
+        result.summary.push(
+          `Associating application currentProcessingAssessment as assessment id ${nextAssessment.id}.`,
+        );
+        const applicationUpdateResult = await entityManager
+          .getRepository(Application)
+          .update(application.id, {
+            currentProcessingAssessment: { id: nextAssessment.id },
+          });
+        if (!applicationUpdateResult.affected) {
+          throw new Error("Application update did not affected any records.");
```

Comment:
The error message could be more specific e.g. currentProcessingAssessment id was already set or incorrect application id. something like that.

Reply:
That log itself may not provide the entire picture, but please check below the logs around it. As shown below there is some additional info that would make it easier to have it traced.

About the suggestions:
- "currentProcessingAssessment id was already set" would not be accurate.
- "incorrect application id" can be checked using the additional logs also.

Queueing next pending assessment for application id 123.
Found 2 pending assessment(s). Queueing assessment 999.
Associating application currentProcessingAssessment as assessment id 999.
(...)
Application update did not affected any records.

Reply:
By the way, planning to do some logging improvement in the next PR.